### PR TITLE
21777 add exclude future effective filing flag

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#21777

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 1]
```diff
@@ -47,15 +47,15 @@ class EligibilityDetails:
 
     @classmethod
     def check_business_eligibility(
-        cls, identifier: str, exclude_in_dissolution=True
+        cls, identifier: str, exclude_in_dissolution=True, exclude_future_effective_filing=False
```
Comment (argush3) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:50]: we had talked about things starting to get messy with the exclude parameters and to use an enum to encapsulate `exclude_in_dissolution` as well as `exclude_future_effective_filing`.  

Can we do that?

**Comment by argush3** [comment]
Can we add relevant tests to verify that dissolution related warnings are able to return even when there is a future effective filing?  

We'll want to test change of address for sure as this was what broke the warnings code.

**Comment by kzdev420** [comment]
> Can we add relevant tests to verify that dissolution related warnings are able to return even when there is a future effective filing?
> 
> We'll want to test change of address for sure as this was what broke the warnings code.

Yes, I will add the test too

File: legal-api/src/legal_api/services/involuntary_dissolution.py (hunk 1 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/involuntary_dissolution.py]: I did

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 2]
```diff
@@ -45,18 +45,26 @@ class EligibilityDetails:
...
+    class ExclusionSettings:
+        """Details about the exclude of a business for dissolution."""
+
+        exclude_in_dissolution: bool = True
+        exclude_future_effective_filing: bool = False
+
     @classmethod
     def check_business_eligibility(
-        cls, identifier: str, exclude_in_dissolution=True
+        cls, identifier: str, exclude_settings: ExclusionSettings = ExclusionSettings()
```
Comment (chenhongjing) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:57]: This change also breaks [furnishings job](https://github.com/bcgov/lear/blob/main/jobs/furnishings/src/furnishings/stage_processors/stage_one.py#L55). Can you also modify the job if it's convenient for you?

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 3]
```diff
@@ -94,7 +102,7 @@ def get_in_dissolution_batch_processing(business_id: int):
             one_or_none()
 
     @staticmethod
-    def _get_businesses_eligible_query(exclude_in_dissolution=True):
+    def _get_businesses_eligible_query(exclude_in_dissolution=True, exclude_future_effective_filing=False):
         """Return SQLAlchemy clause for fetching businesses eligible for involuntary dissolution.
 
         Args:
```
Comment (eason-pan-bc) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:108]: Maybe consider updating the Args part of the docString for the newly added `exclude_future_effective_filing` argument?

File: legal-api/src/legal_api/services/involuntary_dissolution.py (hunk 2 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:57]: Yes, It will break. If this change is good, then, I will update other services too

Comment (argush3) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:57]: The updates for other BE services can be done in a separate PR so we can verify if tests pass.

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 4]
```diff
@@ -45,18 +45,26 @@ class EligibilityDetails:
         ar_overdue: bool
         transition_overdue: bool
 
+    @dataclass
+    class ExclusionSettings:
```
Comment (argush3) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:49]: Can we rename this to `EligibilityFilters`?

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 5]
```diff
@@ -45,18 +45,26 @@ class EligibilityDetails:
...
 
         Returns:
             eligible (bool): True if the business is eligible for dissolution.
             eligibility_details (EligibilityDetails): Details regarding eligibility.
         """
-        query = cls._get_businesses_eligible_query(exclude_in_dissolution).\
-            filter(Business.identifier == identifier)
+        query = cls._get_businesses_eligible_query(
+            exclude_settings.exclude_in_dissolution, exclude_settings.exclude_future_effective_filing
+        ).filter(Business.identifier == identifier)
```
Comment (argush3) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:67]: can we just pass the new dataclass(`EligibilityFilters`) for all functions?

File: legal-api/src/legal_api/services/involuntary_dissolution.py [hunk 6]
```diff
@@ -94,7 +102,7 @@ def get_in_dissolution_batch_processing(business_id: int):
             one_or_none()
 
     @staticmethod
-    def _get_businesses_eligible_query(exclude_in_dissolution=True):
+    def _get_businesses_eligible_query(exclude_in_dissolution=True, exclude_future_effective_filing=False):
```
Comment (argush3) [inline legal-api/src/legal_api/services/involuntary_dissolution.py:105]: can we just use the new dataclass(`EligibilityFilters`)?

Comment (kzdev420) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:105]: Updated

File: legal-api/src/legal_api/services/involuntary_dissolution.py (hunk 5 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/involuntary_dissolution.py:67]: Updated

**Review by argush3 — APPROVED** [review]

**Review by chenhongjing — APPROVED** [review]

**Pull request merged**